src/coreclr/src/debug/createdump/crashinfomac.cpp

```diff
@@ -31,6 +31,45 @@ CrashInfo::CleanupAndResumeProcess()
     }
 }
 
+static
+kern_return_t
+SuspendMachThread(thread_act_t thread, int tid)
+{
+    kern_return_t result;
+
+    while (true)
+    {
+        result = thread_suspend(thread);
+        if (result != KERN_SUCCESS)
+        {
+            fprintf(stderr, "thread_suspend(%d) FAILED %x %s\n", tid, result, mach_error_string(result));
+            break;
+        }
+
+        // Ensure that if the thread was running in the kernel, the kernel operation
+        // is safely aborted so that it can be restarted later.
+        result = thread_abort_safely(thread);
+        if (result == KERN_SUCCESS)
+        {
+            break;
+        }
+        else
+        {
+            TRACE("thread_abort_safely(%d) FAILED %x %s\n", tid, result, mach_error_string(result));
+        }
+        // The thread was running in the kernel executing a non-atomic operation
+        // that cannot be restarted, so we need to resume the thread and retry
```

Comment:
For this, can we check if it's actually that operation by checking for `KERN_FAILURE`.

Reply:
Do you mean reverse the condition and check for KERN_FAILURE?

Reply:
At least the little documentation I found was that `KERN_FAILURE` is returned from non-interruptible tries to interrupt the thread.